src/odemis/gui/comp/miccanvas.py

```diff
@@ -1165,6 +1172,182 @@ def update_drawing(self):
+        self.abilities |= {CAN_DRAG}
+        self._data_buffer = None  # numpy Array with the plot data
+        self.display_xrange = None  # range tuple of floats
+        self.display_yrange = None  # range tuple of floats
+        self.data_xrange = None  # range tuple of floats
+        self.data_yrange = None  # range tuple of floats
+
+    def set_data(self, data, unit_x=None, unit_y=None, range_x=None, range_y=None):
+        xs, ys = zip(*data)
+        self.set_1d_data(xs, ys, unit_x, unit_y, range_x, range_y)
+        self.refresh_plot()
+
+    def set_1d_data(self, xs, ys, unit_x=None, unit_y=None, range_x=None, range_y=None):
+        """
+        Note - this function will not plot any data until a call to refresh_plot is made
+        """
+
+        if len(xs) != len(ys):
+            msg = "X and Y list are of unequal length. X: %s, Y: %s, Xs: %s..."
+            raise ValueError(msg % (len(xs), len(ys), str(xs)[:30]))
+        self._data_buffer = (numpy.array(xs), numpy.array(ys))
+        self.unit_x = unit_x
+        self.unit_y = unit_y
+        
+        if range_x is None or range_y is None:
+            _, range_x, _, range_y = self._calc_data_characteristics(zip(xs, ys))
+
+        self.data_xrange = range_x
+        self.data_yrange = range_y
+        self.display_xrange = range_x
+        self.display_yrange = range_y
+        
+    def fit_view_to_content(self, recenter=None):
+        # note - need to do this via an event to the viewport. Otherwise it doesn't
+        # set the axes properly
+        evt = evtFitViewToContent()
+        wx.PostEvent(self, evt)
+
+    def reset_ranges(self):
+        if self._data_buffer is None:
+            return
+
+        self.set_ranges(self.data_xrange, self.data_yrange)
+
+    def refresh_plot(self):
+        """
+        Refresh the displayed data in the plot.
+        """
+        if self._data_buffer is None:
+            return
+
+        (xst , yst) = self._data_buffer
+
+        # Deal with x
+        lo, hi = self.display_xrange
+
+        # w = numpy.where((xst > lo) * (xst < hi))
+        lox = numpy.searchsorted(xst, lo, side="left")
+        hix = numpy.searchsorted(xst, hi, side="right")
+
+        if lox != hix:
+            xs = xst[lox:hix]
+            ys = yst[lox:hix]
+        else:
+            xs = [(hi - lo) / 2 + lo]
+            ys = [yst[lox]]
+
+        # Add a few points onto the beginning and end of the array
+        # to prevent gaps in teh data from appearing
+        if lox > 0 and xs[0] != lo:
+            xs = numpy.insert(xs, 0, xst[lox - 1])
+            ys = numpy.insert(ys, 0, yst[lox - 1])
+
+        if hix < len(xst) and xs[-1] != hi:
+            xs = numpy.append(xs, xst[hix])
+            ys = numpy.append(ys, yst[hix])
+
+        # Re-sample data based on window size
+        if len(xs) > self.ClientSize.x * 10:
+            rs_factor = len(xs) // self.ClientSize.x // 7
```

Comment:
I'd suggest to move this entire block into a separate function, so that it just looks like:
`xs, ys = self._resample_plot(xs, ys, self.ClientSize.x * 7)`
(The 3rd argument is the new approximate length).